### PR TITLE
feat: introduce concept of "storage class" wtih separate dataset for "blob" storage data

### LINKED ISSUE #3029: Support adding data to blob fields
The overall approach is described in #2263 

When an add operation contains blob fields we will:

 * Add the normal fields, as usual
 * Add the blob fields to the remote dataset and obtain a new version
 * Update the master dataset to point to the new version of the remote dataset

The first two steps can happen in parallel.

If the new data does not contain blob fields then the write will happen as normal (these blobs will be set to NULL).

The new data is not allowed to be only blob fields (blob fields can only be added after-the-fact using a merge-insert style insert)

### PR DESCRIPTION
**Pull request by westonpace**
I'm open to terms other than "storage class" as well.  I also think we might want to use something other than "blob" as it can be easy to confuse the concepts of "blob encoding" (e.g. low page overhead, file-like API) and "blob storage class" (fewer rows per file).  In fact, the thresholds can even be different with the blob encoding threshold being 8-16MB but the blob storage class threshold being closer to 128KB

Closes #3029

### CODE REVIEW

File: protos/file.proto [hunk 1]
```diff
@@ -173,4 +173,17 @@ message Field {
...
   map<string, bytes> metadata = 10;
+
+  /// The storage class of the field
+  ///
+  /// This determines the rate at which the field is compacted.
+  ///
+  /// Currently, there are only two storage classes:
+  /// 
+  /// "" - The default storage class.
+  /// "blob" - The field is compacted into fewer rows per fragment.
```
Comment (wjones127) [inline protos/file.proto:184]: is the criteria for this entirely based on size of row values? or do you imagine there will eventually be other considerations?

**Review by wjones127 — APPROVED** [review]
Seems good. I think I'm fine with storage class, though not sure what other values we will want there.

File: protos/file.proto (hunk 1 above)
Comment (westonpace) [reply protos/file.proto:184]: Size is the only thing I can think of and I don't expect us to need more than two storage classes for the next year or two.

It's possible we have a special "foreign" storage class or something like that for values stored in remote datasets but I think that would be best solved in a different way.

At the moment the only other class I can really think of is for extremely large values where file-per-value makes more sense.

I also hadn't quite done the math yet.  We want to balance:
* We can fit 8MiB worth of values into a file (min bound on num_values)
* We don't have more than 10GiB of values in a file (max bound on num_values)

For example, if we start with int32 we need at least 2Mi values (ideally more actually due to compression).  Given 10GiB this gives us a surprisingly small max size of 5KiB (anything larger will occupy more than 10GiB for 2Mi values).

If we are generous and make the cutoff at 64KiB then we need to make sure we have at least 128 values per file which should allow us up to 160MIB per value.

If we find that we can't be generous and we really do want the cutoff at...say...8KiB then we need at least 1024 values per file which means no more than 10MiB per value which is starting to get to the point where we might want three storage classes.

Comment (westonpace) [reply protos/file.proto:184]: > If we find that we can't be generous and we really do want the cutoff at...say...8KiB

This logic is based on the 10GiB limit which is a somewhat arbitrary "let's pretend we have at most 100 or so 'big' columns and we want to keep files under 1TiB" criteria which seems a little conservative.  At 64KiB we're fine as long as we only have 1 or 2 big columns or can tolerate files over 1TiB.

**Pull request merged**